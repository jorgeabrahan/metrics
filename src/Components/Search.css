.search {
  position: sticky;
  top: 3.5rem;
  width: 100%;
  display: flex;
  justify-content: flex-end;
  padding-inline: 0.5rem;
}

.search__input,
.search__submit {
  padding: 0.4rem;
}

.search__input {
  padding-inline: 1rem;
  width: 100%;
  border: none;
  outline: none;
  background-color: var(--ruber);
  color: white;
  border-radius: 5px 0 0 5px;
}

.search__submit {
  border: none;
  outline: none;
  background: none;
  background-color: #6a0b2e;
  border-radius: 0 5px 5px 0;
}

.search__submit span {
  font-size: 1.8rem;
  color: white;
}
