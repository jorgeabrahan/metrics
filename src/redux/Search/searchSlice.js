/* eslint-disable import/no-extraneous-dependencies */
import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  isSearching: false,
};

const searchSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    toggleSearch: (state) => ({
      ...state, isSearching: !state.isSearching,
    }),
  },
});

export const { toggleSearch } = searchSlice.actions;

export default searchSlice.reducer;
