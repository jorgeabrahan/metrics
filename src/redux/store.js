/* eslint-disable import/no-extraneous-dependencies */
import { configureStore } from '@reduxjs/toolkit';
import currenciesSlice from './Currencies/currenciesSlice';

const store = configureStore({
  reducer: {
    currencies: currenciesSlice,
  },
});

export default store;
