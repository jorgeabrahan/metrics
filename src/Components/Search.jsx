/* eslint-disable import/no-extraneous-dependencies */
import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { searchForCurrency } from '../redux/Currencies/currenciesSlice';
import { searchForConversion } from '../redux/Details/detailsSlice';
import { setQuery } from '../redux/Search/searchSlice';
import './Search.css';

const Search = () => {
  const dispatch = useDispatch();
  const searchInput = useRef();
  const { pathname } = useLocation();
  const { query } = useSelector((store) => store.search);

  useEffect(() => {
    searchInput.current.focus();
    return () => {
      dispatch(setQuery(''));
    };
  }, [dispatch]);

  const onSubmit = (e) => {
    e.preventDefault();
    searchInput.current.focus();
    if (pathname.includes('details')) {
      dispatch(searchForConversion(query.toLowerCase()));
      return;
    }
    dispatch(searchForCurrency(query.toLowerCase()));
  };

  return (
    <form className="search" onSubmit={onSubmit}>
      <input
        className="search__input"
        ref={searchInput}
        type="search"
        placeholder="Search for a currency..."
        value={query}
        onChange={({ target }) => dispatch(setQuery(target.value))}
      />
      <button className="search__submit" type="submit">
        <span className="material-symbols-outlined">
          search_check
        </span>
      </button>
    </form>
  );
};

export default Search;
