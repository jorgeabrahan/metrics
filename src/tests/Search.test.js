import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Search from '../Components/Search';
import store from '../redux/store';
import wrap from './testResources';
import { setQuery } from '../redux/Search/searchSlice';

describe('Tests for <Search /> component', () => {
  render(wrap(<Search />));
  it('Should render the search form', () => {
    expect(screen.getByRole('button', { name: 'search_check' })).not.toBeNull();
    expect(screen.getByRole('searchbox')).not.toBeNull(); // search box input
  });

  it('Should have an empty query by default', () => {
    expect(store.getState().search.query).toBe('');
  });

  it('Should change the query when searching', () => {
    store.dispatch(setQuery('USD'));
    expect(store.getState().search.query).toBe('USD');
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrap(<Search />)).toJSON()).toMatchSnapshot();
  });
});
