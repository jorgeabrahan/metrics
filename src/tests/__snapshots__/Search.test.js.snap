// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Search /> component Should match the snapshot 1`] = `
<form
  className="search"
  onSubmit={[Function]}
>
  <input
    className="search__input"
    onChange={[Function]}
    placeholder="Search for a currency..."
    type="search"
    value="USD"
  />
  <button
    className="search__submit"
    type="submit"
  >
    <span
      className="material-symbols-outlined"
    >
      search_check
    </span>
  </button>
</form>
`;
